File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-9.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.ans = 0
        self.traverse = 0
        def dfs(curr):
            if curr is None:
                return
            dfs(curr.left)
            self.traverse += 1
            if self.traverse == k:
                self.ans = curr.val
                return
            dfs(curr.right)
        dfs(root)
        return self.ans

File: Data Structures & Algorithms/valid-binary-search-tree/submission-9.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isGood(root, -1001, 1001);
    }
    bool isGood(TreeNode* root, int minVal, int maxVal) {
        if (!root) {
            return true;
        }
        if (root->val <= minVal or root->val >= maxVal) {
            return false;
        }
        return isGood(root->left, minVal, root->val) and isGood(root->right, root->val, maxVal);
    }
};
